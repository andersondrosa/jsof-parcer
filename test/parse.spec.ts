import JSOF from "../dist";
JSOF.silence = true;

test("it should work", () => {
  //
  const str = `{
    // Comment label
    "label": "My functional json",
    // Comment function
    "function": myCustomFunction(props.foo, props.bar),
    "array": [1, 2, 3, [1, 2, 3, true, false, null,],],
    "nested": {
      /** 
       * Other comment
      */
      "foo": { // comment here
        "bar": baz("lorem", ipsum.dolor([1,2,3,],),),
      },
    },
  }`;

  expect(JSOF.parse(str)).toMatchSnapshot();
});

test("it should break with double comma into object", () => {
  const str = `{ "foo": "bar",, }`;
  const t = () => JSOF.parse(str);
  expect(t).toThrow(Error);
});

test("it should break with double comma into array", () => {
  const str = `["foo","bar",,]`;
  const t = () => JSOF.parse(str);
  expect(t).toThrow(Error);
});

test("it should break when don't close brackets", () => {
  const str = `{"foo": {}`;
  const t = () => JSOF.parse(str);
  expect(t).toThrow(Error);
});

test("it should break when don't close brackets 2", () => {
  const str = `["foo", "bar", "baz"`;
  const t = () => JSOF.parse(str);
  expect(t).toThrow(Error);
});

test("it should break when invalid token '+'", () => {
  const str = `{"foo": "bar", "number": +1}`;
  const t = () => JSOF.parse(str);
  expect(t).toThrow(Error);
});

test("it should work when valid token '-'", () => {
  const str = `{ "number": -1}`;
  const t = () => JSOF.parse(str);
  expect(t).not.toThrow(Error);
});

test("it should work when valid token '-' with digit", () => {
  const str = `{ "number": -a}`;
  const t = () => JSOF.parse(str);
  expect(t).toThrow(Error);
});

test("it should work when valid function path", () => {
  const str = `{"fn": fn(true)}`;
  const t = () => JSOF.parse(str);
  expect(t).not.toThrow(Error);
});

test("it should break when invalid function path", () => {
  const str = `{"fn": 1_fn(true)}`;
  const t = () => JSOF.parse(str);
  expect(t).toThrow(Error);
});

test("it should break when function don't close", () => {
  const str = `{"fn": fn(true}`;
  const t = () => JSOF.parse(str);
  expect(t).toThrow(Error);
});
