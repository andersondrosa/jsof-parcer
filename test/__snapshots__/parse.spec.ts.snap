// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it should work 1`] = `
{
  "array": [
    1,
    2,
    3,
    [
      1,
      2,
      3,
      true,
      false,
      null,
    ],
  ],
  "function": {
    "_type": "invoker",
    "args": [
      {
        "_type": "prop",
        "path": "props.foo",
      },
      {
        "_type": "prop",
        "path": "props.bar",
      },
    ],
    "into": {
      "_type": "prop",
      "path": "myCustomFunction",
    },
  },
  "label": "My functional json",
  "nested": {
    "foo": {
      "bar": {
        "_type": "invoker",
        "args": [
          "lorem",
          {
            "_type": "invoker",
            "args": [
              [
                1,
                2,
                3,
              ],
            ],
            "into": {
              "_type": "prop",
              "path": "ipsum.dolor",
            },
          },
        ],
        "into": {
          "_type": "prop",
          "path": "baz",
        },
      },
    },
  },
}
`;
